Documentation updates (#32)
grafana/issue-team-scheduler: README.md

# GitHub Action(s) to help automatically assign issues to the right persons

> This repository contains 2 actions which are useful to someone assigning issues to dedicated engineering teams and distributing (automatically assigning) them based on the availability of individual engineers:

Reviewer: why the `:`?

```suggestion
This repository contains 2 actions which are useful to someone assigning issues to dedicated engineering teams and distributing (automatically assigning) them based on the availability of individual engineers.
```

## Regex labeler

This action allows to assign labels based on regular expressions which are ranked against the issue title and body it's run against. This allows to assign team specific labels by matching issues against these regex's.

### Inputs

| Parameter  | Type   | Required | Default                     | Description                                |
| ---------- | ------ | -------- | --------------------------- | ------------------------------------------ |
| `gh-token` | String | false    | `${{ github.token }}`       | The github token to be used for API access |
| `cfg-path` | String | true     | `.github/regex-labeler.yml` | Path for regex labeler config file         |

### Configuration

An exemplary configuration looks like this: 

```yaml
required_labels:
labels:

>   regex-labeler:
>     matchers:
>       - regex: "regex"

Reviewer: i think this is a confusing example because the regular expression is "regex", which doesn't make much sense, a regular expression doesn't always have the string "regex" in it.

how about another example that's more obvious to understand:

```suggestion
  numeric-labeler:
    matchers:
      - numeric: "[0-9]+"
```

  ic-assignment:
    matchers:
      - regex: "assignment"
        weight: 10
```

> In this example the action would apply to all new issues (as no `required_labels` are set). If the issue title and/or body contains the word "regex" it would be labeled as `regex-labeler` by default. But if the word `assignment` is matched, the label `ic-assignment` would be assigned instead (given the higher weight).

Reviewer: ```suggestion
In this example the action would apply to all new issues (as no `required_labels` are set). If the issue title and/or body contains numbers it would be labeled as `numeric` by default. But if the word `assignment` is matched, the label `ic-assignment` would be assigned instead (given the higher weight).
```

| `required_labels` | List of Strings             | false    | `[]`    | List of labels which are required to run this action. If it's triggered on an issue which doesn't have **all** required labels, it exits without doing something |
| `labels`          | Map op label configurations | true     | `nil`   | Definition of which labels are assigned by which matcher configuration. Only the label with the best matching matchers is assigned                               |

Matcher configuration:

| Parameter  | Type            | Required | Default | Description                                                                                                                                                                                             |
| ---------- | --------------- | -------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `matchers` | List of Matcher | false    | `[]`    | List matchers which are required to assign this label. Each matching matchers increases the likeliness (by weight) the owning label is assigned. At least one matcher needs to match to assign a label. |

Reviewer: ```suggestion
| `matchers` | List of Matcher | false    | `[]`    | List matchers which are required to assign this label. Each matching matcher increases the likeliness (by weight) the owning label is assigned. At least one matcher needs to match to assign a label. |
```

    requireLabel: 
    - "product_b"
    - "product_b-important"
    members:
    - name: manager
      output: "slack-handle"
ignoreLabels:
  - stale
```

#### Root configuration struct

| Parameter      | Type                       | Required | Default | Description                                                                                                                                                          |
| -------------- | -------------------------- | -------- | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `ignoreLabels` | List of Strings            | false    | `[]`    | List of labels which mark this issue to be ignored. If triggered on an issue which has **one** of the labels to be ignored, the action exits without doing something |

Reviewer: ```suggestion
| `ignoreLabels` | List of Strings            | false    | `[]`    | List of labels which mark this issue to be ignored. If triggered on an issue which has at least **one** of the labels to be ignored, the action exits without doing something |
```

| `googleCalendar` | String | false    | ``      | Google Calendar name which is checked through the specified service account to determine availability. If set, `ical-url` is ignored.                                                                                                        |

### Considerations

#### Timezone awareness

During the initial implementation it was carefully discussed if the action should take timzeones into consideration (aka only assign issues to someone during their working hours). After in-depth analysis of issues in other projects we concluded that the creation time of issues don't align to working hours of the majority of the teams we were working in (e.g. majority of issues created during american working hours while majority of the team wasn't located there). **This led us to make the action not timezone aware as of today.** As this obviously cause issues when you need issues to handle in a timely manner we are open to proposals for improvement, if they take into account that our main priority is to distribute issues as fairly as possible across a team. 

#### Fairness

> Fairness in assigning issues to individuals of a team is subjective. As mean time to resolve depends on time of creation, issue complexity, productivity, knowledge, experience, availability and some degree of luck per team member it needs to be acknowledged that their is no objectional fair distribution of issues among a group of people. Taking all these metrics into account is almost impossible and most likely would require additional state which we would need to store somewhere. We also discussed ideas like the tracking the amount of completed issues per IC, but had to realise that issues which were hard to resolve (and therefore take a long time) would suffer from this. We will revisit the current approach after gaining experience with it, but start lightweight now. 

Reviewer: ```suggestion
Fairness in assigning issues to individuals of a team is subjective. As mean time to resolve depends on time of creation, issue complexity, productivity, knowledge, experience, availability and some degree of luck per team member it needs to be acknowledged that their is no objective fair distribution of issues among a group of people. Taking all these metrics into account is almost impossible and most likely would require additional state which we would need to store somewhere. We also discussed ideas like tracking the amount of completed issues per IC, but had to realise that issues which were hard to resolve (and therefore take a long time) would suffer from this. We will revisit the current approach after gaining experience with it, but start lightweight now. 
```
